File: Java/DataStructures/MyCircleQueue.java
package Java.DataStructures;

public class MyCircleQueue {

    private int queue[];
    private int capacity;
    private int front;
    private int rear;

    public MyCircleQueue(int size){
        queue = new int[size];
        capacity = size;
        front = -1;
        rear = -1;
    }

    public void enQueue(int item){

        if(isFull()){
            throw new RuntimeException("Queue is full");
        }

        if(front == -1) front++;

        rear = (rear + 1) % capacity;
        queue[rear] = item;

    }

    public int deQueue(){

        if(isEmpty()){
            throw new RuntimeException("Queue is empty");
        }

        int frontElement = queue[front];
        front = (front + 1) % capacity;

        if(front == rear){
            front = -1;
            rear = -1;
        }

        return frontElement;
    }

    public Boolean isEmpty(){
        return front == rear;
    }

    public Boolean isFull(){
        return (front == rear + 1) || ((rear == capacity - 1) && (front == 0));
    }

    public int peek(){
        
        if(isEmpty()){
            throw new RuntimeException("Queue is empty");
        }

        return queue[front];
    }

    public void printQueue(){
        String queueAsString = "[";

        if(!isEmpty()){
            if(front < rear){
                for(int i = front; i <= rear; i++){
                    if(i == front){
                        queueAsString = queueAsString  + Integer.toString(queue[i]);
                    }else{
                        queueAsString = queueAsString + "| " + Integer.toString(queue[i]);
                    }
                }

            }else{
                for(int i = front; i < capacity; i++){
                    if(i == capacity - 1){
                        queueAsString = queueAsString + Integer.toString(queue[i]);
                    }else{
                        queueAsString = queueAsString + Integer.toString(queue[i]) + "| ";
                    }
                }

                for(int i = 0; i <= rear; i++){
                    queueAsString = queueAsString + "| " + Integer.toString(queue[i]);
                }
            }
        }

        queueAsString = queueAsString + "] <- Rear";
        System.out.println(queueAsString);
    }

}

File: Python/Data Structures/MyCircleQueue.py
class MyCircleQueue:

    def __init__(self, size):
        self.capacity = size
        self.front = -1
        self.rear = -1
        self.queue = [None] * size
    
    def enQueue(self, item):
        
        if self.isFull():
            raise Exception("Queue is Full.")

        if self.front == -1:
            self.front = self.front + 1
        
        self.rear = (self.rear + 1) % self.capacity

        self.queue[self.rear] = item

    def deQueue(self):
        
        if self.isEmpty():
            raise Exception("Queue is Empty.")
        
        frontElement = self.queue[self.front]
        self.front = (self.front + 1) % self.capacity

        if self.front == self.rear:
            self.rear = -1
            self.front = -1

        return frontElement

    def isEmpty(self):
        return self.rear == self.front

    def isFull(self):
        return (self.front == self.rear + 1) or ((self.rear == self.capacity - 1) and (self.front == 0))

    def peek(self):
        
        if self.isEmpty():
            raise Exception("Queue is Empty.")
        
        return self.queue[self.front]

    def __str__(self):
        
        if self.isEmpty():
            return '[]'
        
        if self.front < self.rear:
            return f'{[self.queue[x] for x in range(self.front, self.rear + 1)]} <- Rear'
        
        frontElements = "["
        rearElements = ""
        for x in range(self.front, self.capacity):
            if x == self.capacity - 1:
                frontElements = frontElements + str(self.queue[x])
            else:
                frontElements = frontElements + str(self.queue[x]) + ", "

        for x in range(0, self.rear + 1):
            rearElements = rearElements + ", " + str(self.queue[x])
        

        return frontElements + rearElements + "] <- Rear"

